src/3d/qgscolorramptexture.cpp

```diff
@@ -0,0 +1,117 @@
+  : mColorRampShader( colorRampShader )
+  , mVerticalScale( verticalScale )
+{
+}
+
+Qt3DRender::QTextureImageDataPtr QgsColorRampTextureGenerator::operator()()
+{
+  Qt3DRender::QTextureImageDataPtr dataPtr = Qt3DRender::QTextureImageDataPtr::create();
+  dataPtr->setFormat( QOpenGLTexture::RGBA32F );
+  dataPtr->setTarget( QOpenGLTexture::Target1D );
+  dataPtr->setPixelFormat( QOpenGLTexture::RGBA );
+  dataPtr->setPixelType( QOpenGLTexture::Float32 );
+
+  QByteArray data;
+  QList<QgsColorRampShader::ColorRampItem> colorItemList = mColorRampShader.colorRampItemList();
+  int size = colorItemList.count() ;
+
+  dataPtr->setWidth( size );
+  dataPtr->setHeight( 1 );
+  dataPtr->setDepth( 1 );
+  dataPtr->setFaces( 1 );
+  dataPtr->setLayers( 1 );
+  dataPtr->setMipLevels( 1 );
+
+  for ( int i = 0; i < colorItemList.count(); ++i )
+  {
+    float mag = float( colorItemList.at( i ).value * mVerticalScale );
+
+    QColor color = colorItemList.at( i ).color;
+    float rf = float( color.redF() );
+    float gf = float( color.greenF() );
+    float bf = float( color.blueF() );
+
+    data.append( reinterpret_cast<const char *>( &mag ), sizeof( float ) );
+    data.append( reinterpret_cast<const char *>( &rf ), sizeof( float ) );
+    data.append( reinterpret_cast<const char *>( &gf ), sizeof( float ) );
+    data.append( reinterpret_cast<const char *>( &bf ), sizeof( float ) );
+
+  }
+
+  dataPtr->setData( data, sizeof( float ) ); //size is the size of the type, here float
+
+  return dataPtr;
+}
+
+bool QgsColorRampTextureGenerator::operator==( const Qt3DRender::QTextureImageDataGenerator &other ) const
+{
+  const QgsColorRampTextureGenerator *otherFunctor = functor_cast<QgsColorRampTextureGenerator>( &other );
+  if ( !otherFunctor )
+    return false;
+
+  QgsColorRampShader otherColorRampShader = otherFunctor->mColorRampShader;
+
+  if ( mColorRampShader.colorRampItemList().count() != otherColorRampShader.colorRampItemList().count() ||
+       mColorRampShader.classificationMode() != otherColorRampShader.classificationMode() ||
+       mColorRampShader.colorRampType() != otherColorRampShader.colorRampType() )
+  {
+    return false;
+  }
+
+  QList<QgsColorRampShader::ColorRampItem> colorItemList = mColorRampShader.colorRampItemList();
+  QList<QgsColorRampShader::ColorRampItem> otherColorItemList = otherColorRampShader.colorRampItemList();
+  for ( int i = 0; i < colorItemList.count(); ++i )
+  {
+    const QColor color = colorItemList.at( i ).color;
+    const QColor otherColor = otherColorItemList.at( i ).color;
+    double value = colorItemList.at( i ).value;
+    double otherValue = otherColorItemList.at( i ).value;
+    if ( color != otherColor ||
+         ( !std::isnan( value ) && !std::isnan( otherValue ) && colorItemList.at( i ).value != otherColorItemList.at( i ).value ) ||
+         ( std::isnan( value ) != std::isnan( otherValue ) ) )
+      return false;
+  }
```

Comment:
It would be good to move all this code to QgsColorRampShader class - and implement its `bool operator==( const QgsColorRampShader &other)` method and `operator!=` (which would just use the equals operator) - then here we can just use `return mColorRampShader == otherFunctor->mColorRampShader` and the code would be easier to read.

Also, it would be good to implement `operator==` in QgsColorRampShader::ColorRampItem, so then in QgsColorRampShader you can simply compare the two lists for equality.